### DISCUSSION TITLE
Intelligent Slots Displayed

### DISCUSSION THREAD

**Discussion by liquidv** [BottlecapDave/HomeAssistant-OctopusEnergy#500] [General]
Hello, is there anyway to show the Intelligent slot hours on HA like this but using this integration rather than using Megakids Intelligent integration? I'm running both at the moment and it's a bit fiddly as I want both gas tracker price from this integration, and OI slot hours from the other one.

Megakids intelligent integration appears to go to an unknown state a bit too often, whereas this integration appears more stable hence I want to stick with this one only.

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/122752718/c792cfca-38d4-4b7c-972a-5726df02a9c4)

**Comment by Dave-Parry** [comment]
I also am interested in displaying the slots on my dashboard, but no idea how to access the data

**Reply by BottlecapDave** [reply]
Hi, sorry for the late response. The same data is available as attributes on the [binary_sensor.octopus_energy_intelligent_dispatching](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/intelligent.md#is-dispatching) sensor. If you can supply a link to where you got the screenshot, I might be able to help.

**Reply by liquidv** [reply]
The code is here if you need it for the other Octopus Intelligent integration.

https://community.home-assistant.io/t/octopus-energy-intelligent-tariff/418680/123

**Reply by Dave-Parry** [reply]
Thanks for reply, but I'm on agile not intelligent Tariff so don't have that range of sensors

**Comment by liquidv** [comment]
> Hi, sorry for the late response. The same data is available as attributes on the [binary_sensor.octopus_energy_intelligent_dispatching](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/intelligent.md#is-dispatching) sensor. If you can supply a link to where you got the screenshot, I might be able to help.

https://community.home-assistant.io/t/octopus-energy-intelligent-tariff/418680/123

{% if state_attr('binary_sensor.octopus_intelligent_slot', 'plannedDispatches') | length > 0 %}
{% for dispatch in state_attr('binary_sensor.octopus_intelligent_slot', 'plannedDispatches') -%}
⏰ {{ (dispatch.startDtUtc | as_datetime | as_local).strftime('%H:%M') }} to {{ (dispatch.endDtUtc | as_datetime | as_local).strftime('%H:%M') }}
{{ "\n" }}
{%- endfor %}
{%- else -%}
⏰ None
{%- endif %}

**Reply by BottlecapDave** [reply]
This can be adjusted to look at the sensor for this integration like so
```
{% if state_attr('binary_sensor.octopus_energy_intelligent_dispatching', 'planned_dispatches') | length > 0 %}
{% for dispatch in state_attr('binary_sensor.octopus_energy_intelligent_dispatching', 'planned_dispatches') -%}
⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
{{ "\n" }}
{%- endfor %}
{%- else -%}
⏰ None
{%- endif %}
```

**Reply by liquidv** [reply]
hmm doesn't seem to be working.....

**Reply by BottlecapDave** [reply]
I've updated the above example which should now work

**Reply by liquidv** [reply]
Brilliant it works, thank you!
Now I can uninstall the other Intelligent integration and just use your one. Cheers

**Comment by reedy** [comment]
I came here looking for the same thing, and putting the instructions together (I'll move it to one of the docs files when I've got it working)...

Install https://github.com/thomasloven/lovelace-template-entity-row

Add this to your card

```yaml
  - type: custom:template-entity-row
    name: Planned slots
    secondary: >-
      {% if state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching', 'planned_dispatches') | length > 0 %}
        {% for dispatch in state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching', 'planned_dispatches') -%}
          ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
           {{ "" }}
        {%- endfor %}
      {%- else -%}
        ⏰ None
      {%- endif %}
```

replacing `XXXXXXXX` as appropriate...

Which is currently giving

![Screenshot 2023-11-22 at 20 22 54](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/67615/02268a34-e0ee-478e-963a-f5ae795a4d33)

The lack of newlines is a bit annoying, but a limitation of the card, seemingly - https://community.home-assistant.io/t/octopus-energy-intelligent-tariff/418680/108 / https://github.com/thomasloven/lovelace-template-entity-row/issues/71 / https://github.com/thomasloven/lovelace-template-entity-row/pull/84

Though later comments in the thread make it sound like it works for them...

I'd like to make the "off" disappear too... But that seems it's a different known bug, ala https://github.com/thomasloven/lovelace-template-entity-row/issues/106

So setting `state: ''` fixes it. Also adding an icon to make it look a bit more consistent:

```yaml
type: custom:template-entity-row
name: Planned slots
state: ''
icon: mdi:clock
secondary: >-
  {% if
  state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching',
  'planned_dispatches') | length > 0 %}
    {% for dispatch in state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching', 'planned_dispatches') -%}
      ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
      {{ "\n" }}
    {%- endfor %}
  {%- else -%}
    ⏰ None
  {%- endif %}
```

![Screenshot 2023-11-22 at 20 50 04](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/67615/32596f0f-84c0-41f4-97c3-8159b36a2210)

**Reply by BottlecapDave** [reply]
You can also use the built in markdown card, which might support new lines (I'm not on intelligent myself)

**Reply by reedy** [reply]
```yaml
type: markdown
title: Planned slots
content: |-
  {% if
  state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching',
  'planned_dispatches') | length > 0 %}
    {% for dispatch in state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching', 'planned_dispatches') -%}
      ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
      {{ "\n" }}
    {%- endfor %}
  {%- else -%}
    ⏰ None
  {%- endif %}
```

gives

![Screenshot 2023-11-22 at 21 02 55](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/67615/4338ab22-64f9-476e-8367-8b2c114490c0)

**Reply by reedy** [reply]
If we install https://github.com/thomasloven/lovelace-hui-element too...

```
  - type: custom:hui-element
    card_type: markdown
    title: Planned slots
    content: |-
      {% if
      state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching',
      'planned_dispatches') | length > 0 %}
        {% for dispatch in state_attr('binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching', 'planned_dispatches') -%}
          ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
          {{ "\n" }}
        {%- endfor %}
      {%- else -%}
        ⏰ None
      {%- endif %}
```

![Screenshot 2023-11-22 at 21 18 29](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/67615/dc64cb39-6564-4044-8a01-36355aaa1262)

**Reply by neildsb** [reply]
awesome just what I was looking for, top job!!

**Reply by neildsb** [reply]
any ideas how to integrate an availability state on bootup to stop a log error?

like -
availability: "{{ (is_state('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching','off' )) or (is_state('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching','on' )) }}"

Error while processing template: Template<template=({% if state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') | length > 0 %} {% for dispatch in state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') -%} ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} {{ "\n" }} {%- endfor %} {%- else -%} ⏰ None {%- endif %}) renders=2>
Error while processing template: Template<template=({% if state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') | length > 0 %} {% for dispatch in state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') -%} ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} {{ "\n" }} {%- endfor %} {%- else -%} ⏰ None {%- endif %}) renders=2>
Error while processing template: Template<template=({% if state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') | length > 0 %} {% for dispatch in state_attr('binary_sensor.octopus_energy_a_xxxx_intelligent_dispatching', 'planned_dispatches') -%} ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} {{ "\n" }} {%- endfor %} {%- else -%} ⏰ None {%- endif %}) renders=6>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 585, in async_render
    render_result = _render_with_context(self.template, compiled, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2600, in _render_with_context
    return template.render(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.12/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: object of type 'NoneType' has no len()

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 713, in async_render_to_info
    render_info._result = self.async_render(
                          ^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 587, in async_render
    raise TemplateError(err) from err
homeassistant.exceptions.TemplateError: TypeError: object of type 'NoneType' has no len()

<img width="289" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/91122872/997cf71b-165b-41ad-9e0d-bba67589f0ac">

**Comment by Dave-Parry** [comment]
Can I ask how to get "binary_sensor.octopus_energy_a_XXXXXXXX_intelligent_dispatching" as I don't have any dispatching sensors.

**Reply by BottlecapDave** [reply]
You have to have an active Octopus intelligent tariff. If you have just switched tariffs, you need to restart your HA instance or reload the integration to get the new sensors. If you have just switched, it can sometimes take a few days for the API to report the new tariff.

**Comment by Dave-Parry** [comment]
No on agile, ok so that's why I don't have them. Shame would have liked to use this.

**Reply by CraigCallender** [reply]
@Dave-Parry, if you're still in need, something like this is how I get the charge slots I've configured for my home battery via this integration:

```
{% if
state_attr('binary_sensor.octopus_energy_target_battery_charge_slots',
'target_times') | length > 0 %}
  {% for dispatch in state_attr('binary_sensor.octopus_energy_target_battery_charge_slots', 'target_times') -%}
    ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} @ {{ '%0.2f' % (dispatch.value_inc_vat*100) }}p/kWh
    {{ "\n" }}
  {%- endfor %}
{%- else -%}
  ⏰ None
{%- endif %}
```
Which gives this:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/125502786/c59e5ce3-dfbf-4fb6-82c6-d1faa2f7f9ee)


The `binary_sensor.octopus_energy_battery_charge_slots` is the name a target rate I've configured via the Octopus Energy HA Integration.

**Comment by Ron-ski** [comment]
I've tried to implement this, I do get the list, but I get no title.

I've tried adding title: Planned Slots but this gets moved to the end, and doesn't appear, also tried adding label: Planned slots, still nothing.

Any ideas please?

```
type: custom:hui-element
card_type: markdown
content: |-
  {% if
  state_attr('binary_sensor.octopus_energy_a_XXXXXX_intelligent_dispatching',
  'planned_dispatches') | length > 0 %}
    {% for dispatch in state_attr('binary_sensor.octopus_energy_a_XXXXXX_intelligent_dispatching', 'planned_dispatches') -%}
      ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
      {{ "\n" }}
    {%- endfor %}
  {%- else -%}
    ⏰ None
  {%- endif %}
cards:
  - type: heading
    heading: Planned Slots
```

**Reply by reedy** [reply]
Use ```  (line before and line at the end) for blocks of code :)

What if you just have a title? As per https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/500#discussioncomment-7646395

Trying to add a `type: heading` in place, my HA doesn't like it... It needs to be under a card etc...

So you may need to post your code with the wider context for it to work...

**Reply by Ron-ski** [reply]
Thanks, I originally copied and pasted reply 500, but it moves the title bit to the end as per my code above, and I don't get a title.

Not sure what you mean by  "It needs to be under a card etc..." or you're last sentences, sorry but I'm not overly familiar with HA, or laying out cards.

Well, bizarrely,  I moved the code to an existing section, but a new card, added the title and it works.

Previously it was a new section, I'm guessing each section has cards, and that's what you meant by under a card, stuff needs to be in cards, and cards reside in section.

Thanks

**Reply by reedy** [reply]
I can't say I'm particularly familiar with them either... Just know that there's a hierarchy and that some stuff can go under others, but then not under some others.

Glad you figured it out though!

Depending on how you're doing it, it will give you some visible errors in their "visual editor". They're not the most obvious some of the time, but should be pointing you in roughly the right direction.

**Comment by zuhooo** [comment]
Just did some research, I think there was an update to the entity name, so here is the new code I worked out, it displayed “None” when there is no planned dispatch, and will display all the planned dispatches when there are more than 1 elements in the attributes.
I am using lovelace dashboard UI btw.
![image](https://github.com/user-attachments/assets/37ef68a6-5dc7-4f1e-b486-2db8098c2e99)
don’t forget to replace xxxxxxx with your IOG account number
```
{% set dispatches = state_attr('binary_sensor.octopus_energy_xxxxxxxx_intelligent_dispatching', 'planned_dispatches') | default([]) %}

{% if dispatches | length > 0 %}
  {% for dispatch in dispatches %}
## ⏰ {{ as_timestamp(dispatch.start) | timestamp_custom('%I:%M %p') }} -     {{ as_timestamp(dispatch.end) | timestamp_custom('%I:%M %p') }}
  {% endfor %}
{% else %}
## ⏰ None
{% endif %}
```

**Comment by philip98642** [comment]
Hi @BottlecapDave, 
I'm using a modified version of this code to display the completed dispatches in a markdown card: 

```
{% if
  state_attr('binary_sensor.octopus_energy_a_xxxxxx_intelligent_dispatching','completed_dispatches') | length > 0 %}
    {% for dispatch in state_attr('binary_sensor.octopus_energy_a_xxxxxx_intelligent_dispatching', 'completed_dispatches') -%}
    ⏰ {{ (dispatch.start | as_local).strftime('%d/%m/%Y %H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} ⚡️{{ ((dispatch.charge_in_kwh) * -1) }}kWh 
      {{ "\n" }}
    {%- endfor %}
  {%- else -%}
    ⏰ None
{%- endif %}

```
The data is displayed but the values for each hour are displayed in an odd order e.g. 01:00 then 00:30 then 02:00:


```
⏰ 23/03/2025 01:00 to 01:30 ⚡️2.41kWh
⏰ 23/03/2025 00:30 to 01:00 ⚡️0.09kWh
⏰ 23/03/2025 02:00 to 02:30 ⚡️2.37kWh
⏰ 23/03/2025 01:30 to 02:00 ⚡️0.28kWh
⏰ 23/03/2025 03:00 to 03:30 ⚡️1.94kWh
⏰ 23/03/2025 02:30 to 03:00 ⚡️0.26kWh

```
I wondered if it would be easy for you to sort the data by start time as it goes into the attribute as I am having trouble sorting it when I get it out of the attribute? 

P.S. I realise you don't have intelligent Octopus go so if there is some raw data or debugging I can send to help let me know.

**Comment by reedy** [comment]
Replaced the ID obviously... Posting the log for searchability in future

```
Logger: homeassistant.helpers.event
Source: helpers/template.py:643
First occurred: 01:29:53 (2 occurrences)
Last logged: 01:29:53

Error while processing template: Template<template=({% if state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') | length > 0 %} {% for dispatch in state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') -%} ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} {{ "\n" }} {%- endfor %} {%- else -%} ⏰ None {%- endif %}) renders=2>
Error while processing template: Template<template=({% if state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') | length > 0 %} {% for dispatch in state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') -%} ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }} {{ "\n" }} {%- endfor %} {%- else -%} ⏰ None {%- endif %}) renders=6>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 641, in async_render
    render_result = _render_with_context(self.template, compiled, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2977, in _render_with_context
    return template.render(**kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/jinja2/environment.py", line 1295, in render
    self.environment.handle_exception()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/local/lib/python3.13/site-packages/jinja2/environment.py", line 942, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
TypeError: object of type 'NoneType' has no len()

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 758, in async_render_to_info
    render_info._result = self.async_render(  # noqa: SLF001
                          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
        variables, strict=strict, log_fn=log_fn, **kwargs
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 643, in async_render
    raise TemplateError(err) from err
homeassistant.exceptions.TemplateError: TypeError: object of type 'NoneType' has no len()
```

```yaml
  - type: custom:hui-element
    card_type: markdown
    title: Planned slots
    content: >-
      {% if
      state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching',
      'planned_dispatches') | length > 0 %}
        {% for dispatch in state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') -%}
          ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
          {{ "\n" }}
        {%- endfor %}
      {%- else -%}
        ⏰ None
      {%- endif %}
```

into

```yaml
  - type: custom:hui-element
    card_type: markdown
    title: Planned slots
    content: >-
      {% set dispatches = state_attr('binary_sensor.octopus_energy_{{DEVICE_ID}}_intelligent_dispatching', 'planned_dispatches') | default([], true) %}

      {% if dispatches | length > 0 %}
        {% for dispatch in dispatches -%}
          ⏰ {{ (dispatch.start | as_local).strftime('%H:%M') }} to {{ (dispatch.end | as_local).strftime('%H:%M') }}
          {{ "\n" }}
        {%- endfor %}
      {%- else -%}
        ⏰ None
      {%- endif %}
```

from https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/500#discussioncomment-12511287

Yay, de-duplication too!
